File: 力扣/每日一题/107. 二叉树的层次遍历 II.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def levelOrderBottom(self, root: TreeNode):
        if not root:
            return []
        ans_lst = []
        dq = []
        dq.append(root)
        def helper(dq:list):
            if len(dq) == 0:
                return 
            new_dq = []
            tmp_lst = []
            while len(dq) > 0:
                node = dq.pop(0)
                tmp_lst.append(node.val)
                if node.left:
                    new_dq.append(node.left)
                if node.right:
                    new_dq.append(node.right)
            ans_lst.append(tmp_lst)
            helper(new_dq)
        helper(dq)
        return ans_lst[::-1]

File: c++代码/107.二叉树的层次遍历II.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
    vector<vector<int>> levelOrderBottom(TreeNode *root)
    {
        if (!root)
        {
            return v;
        }
        q.push(root);
        this->helper(q);
        reverse(v.begin(), v.end());
        return v;
    }

    void helper(queue<TreeNode *> q)
    {
        if (q.empty()) return;
        vector<int> tmp_v;
        queue<TreeNode *> new_q;
        while (!q.empty())
        {
            TreeNode *node = q.front();
            q.pop();
            tmp_v.push_back(node->val);
            if (node->left)
                new_q.push(node->left);
            if (node->right)
                new_q.push(node->right);
        }
        v.push_back(tmp_v);
        this->helper(new_q);
    }
    queue<TreeNode *> q;
    vector<vector<int>> v;
};
